cmd/lotus-shed/indexes.go

```diff
@@ -879,3 +900,339 @@ var backfillTxHashCmd = &cli.Command{
+		if err != nil {
+			return xerrors.Errorf("error parsing prune args: %w", err)
+		}
+
+		if err := pruneEventsIndex(ctx, basePath, startHeight); err != nil {
+			return xerrors.Errorf("error pruning events index: %w", err)
+		}
+
+		return nil
+	},
+}
+
+// pruneAllIndexesCmd is a command that prunes all the indexes
+var pruneAllIndexesCmd = &cli.Command{
+	Name:  "prune-all-indexes",
+	Usage: "Prune all the indexes DBs (msgindex, txhash, events) older than a given height",
+	Flags: []cli.Flag{
+		&cli.IntFlag{
+			Name:  "older-than",
+			Value: 0,
+			Usage: "height to start pruning events older than this epoch, 0 means start from head",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		srv, err := lcli.GetFullNodeServices(cctx)
+		if err != nil {
+			return err
+		}
+		defer srv.Close() //nolint:errcheck
+
+		api := srv.FullNodeAPI()
+		ctx := lcli.ReqContext(cctx)
+
+		basePath, startHeight, err := parsePruneArgs(ctx, cctx, api)
+		if err != nil {
+			return xerrors.Errorf("error parsing prune args: %w", err)
+		}
+
+		var g = new(errgroup.Group)
+
+		g.Go(func() error {
+			if err := pruneMsgIndex(ctx, basePath, startHeight); err != nil {
+				return xerrors.Errorf("error pruning msgindex: %w", err)
+			}
+
+			return nil
+		})
+
+		g.Go(func() error {
+			if err := pruneTxIndex(ctx, api, basePath, startHeight); err != nil {
+				return xerrors.Errorf("error pruning txindex: %w", err)
+			}
+
+			return nil
+		})
+
+		g.Go(func() error {
+			if err := pruneEventsIndex(ctx, basePath, startHeight); err != nil {
+				return xerrors.Errorf("error pruning events index: %w", err)
+			}
+
+			return nil
+		})
+
+		return g.Wait()
+	},
+}
+
+// pruneEventsIndex is a helper function that prunes the events index.db for a number of epochs starting from a specified height
+func pruneEventsIndex(_ context.Context, basePath string, startHeight int64) error {
+	log.Infof("pruning events index")
+
+	dbPath := path.Join(basePath, databaseType, filter.DefaultDbFilename)
+	db, err := sql.Open(databaseDriver, dbPath+"?_txlock=immediate")
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err := db.Close()
+		if err != nil {
+			fmt.Printf("ERROR: closing db: %s", err)
+		}
+	}()
+
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+
+	rollback := func(tableName string) {
+		if err := tx.Rollback(); err != nil {
+			fmt.Printf("ERROR: rollback %s tx :%s", tableName, err)
+		}
+	}
+
+	res, err := tx.Exec(deleteEventFromStartHeight, startHeight)
+	if err != nil {
+		rollback(eventTable)
+		return xerrors.Errorf("error deleting event: %w", err)
+	}
+
+	var totalRowsAffected int64
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("error getting rows affected: %w", err)
+	}
+
+	totalRowsAffected += rowsAffected
+	log.Infof("pruned %s table, rows affected: %d", eventTable, rowsAffected)
+
+	res, err = tx.Exec(deleteEventEntriesByEventIDs, startHeight)
```

Comment:
Here calling `res.RowsAffected()` is returning 0 even though in DB data is being deleted.

My guess is that SQLite not able to figure out how many rows are affected because there is a subquery inside it.

What are your thoughts on this @rvagg.

Reply:
you just need to swap this with `deleteEventFromStartHeight` - this query relies on the entries being in the `event` table because of the sub-select, but you've just deleted the ones that it's trying to match; so clean up `event_entry` first, then `event` and you're all good

Reply:
As I mentioned the data is still being deleted but, it's not returning the number of rows affected.

Anyway, I will update it as you mentioned.

Reply:
But how do you know it's being deleted? you've lost the entried they join with on the `event` table, have you tried a `select count(*) from event` and `select count(*) from event_index` to see that they're actually deleted? Because of the subquery working on rows that were previously deleted, you should not have the join and therefore not get deletion.

Reply:
You're correct, I was checking it the wrong way.